src/txn/txn.c

```diff
@@ -1393,20 +1393,20 @@ __wt_txn_rollback_required(WT_SESSION_IMPL *session, const char *reason)
 int
 __wt_txn_init(WT_SESSION_IMPL *session, WT_SESSION_IMPL *session_ret)
 {
+    WT_CONNECTION_IMPL *conn;
     WT_TXN *txn;
 
-    txn = &session_ret->txn;
+    conn = S2C(session);
+
+    /* Allocate the WT_TXN structure, including a variable length array of snapshot information. */
+    WT_RET(__wt_calloc(session, 1, sizeof(WT_TXN) + sizeof(txn->snapshot[0]) * conn->session_size,
+      &session_ret->txn));
+    txn = session_ret->txn;
+    txn->snapshot = txn->__snapshot;
```

Comment:
What's the purpose of having two members?

Reply:
This is the trick that allows us to keep the same number of allocations/frees as before.

We were already allocating the snapshot array on transaction begin, and so I changed it to be a variable-length array trailing the `WT_TXN` structure, so instead of allocating the snapshot array at transaction begin, I'm allocating both the `WT_TXN` structure and the snapshot array in a single allocation at transaction begin.